subprojects/plugin-use/src/main/java/org/gradle/plugin/management/internal/PluginResolutions.java

```diff
@@ -0,0 +1,33 @@
+package org.gradle.plugin.management.internal;
+
+import org.gradle.api.Action;
+import org.gradle.api.Nullable;
+import org.gradle.plugin.management.PluginRequest;
+import org.gradle.plugin.management.PluginResolveDetails;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public class PluginResolutions {
+
+    private final Set<Action<? super PluginResolveDetails>> resolutionDetails = new LinkedHashSet<Action<? super PluginResolveDetails>>();
+
+    public void add(Action<? super PluginResolveDetails> rule) {
+        resolutionDetails.add(rule);
+    }
+
+    @Nullable
+    public ResolvedPluginRequest resolveRequest(PluginRequest pluginRequest) {
```

Comment:
This should never return null. If the user doesn't change the original request, then the original request is returned here.